File: Python/Subsets II.py
# Given a collection of integers that might contain duplicates, S, return all possible subsets.

# Note:
# Elements in a subset must be in non-descending order.
# The solution set must not contain duplicate subsets.
# For example,
# If S = [1,2,2], a solution is:

# [
  # [2],
  # [1],
  # [1,2,2],
  # [2,2],
  # [1,2],
  # []
# ]


class Solution:
    # @param num, a list of integer
    # @return a list of lists of integer
	def subsetsWithDup(self, S):
		result = []
		output = []
		if len(S) == 0:
			return result
		result.append(list(output))
		S.sort()
		self.generateSub(S,0,result,output)
		return result
	def generateSub(self,S,step,result,output):
		for i in range(step,len(S)):
			output.append(S[i])
			temp = list(output)
			if temp not in result:
				result.append(temp)
			if i < (len(S) - 1):
				self.generateSub(S,i+1,result,output)
			output.pop()
#
class Solution:
    # @param S, a list of integer
    # @return a list of lists of integer
	def subsetsWithDup(self, S):
		result = []
		output = []
		if len(S) == 0:
			return result
		result.append(list(output))
		S.sort()
		self.generateSub(S,0,result,output)
		return result
	def generateSub(self,S,step,result,output):
		i = step
		while i < len(S):
			output.append(S[i])
			result.append(list(output))
			if i < (len(S) - 1):
				self.generateSub(S,i+1,result,output)
			output.pop()
			while (i<len(S)-1) and S[i]==S[i+1]:
				i = i + 1
			i = i + 1

File: Java/Subsets II.java
class Subsets II{
	public ArrayList<ArrayList<Integer>> subsetsWithDup(int[] num) {
		ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> solution = new ArrayList<Integer>();
        if(num == null || num.length == 0)
            return res;
        Arrays.sort(num);
        findSubsets(solution,res,0,num);
        res.add(new ArrayList<Integer>());
        return findSubsetsIterative(num);
	}
	private void findSubsets(ArrayList<Integer> solution, 
		ArrayList<ArrayList<Integer>> res, int level, int[] num){
		if(level == num.length)
			return;
		for(int i=level;i<num.length;i++){
			if(i<num.length-1&&num[i] == num[i+1])
				continue;
			solution.add(num[i]);
			res.add(new ArrayList<Integer>(solution));
			findSubsets(solution,res,i+1,num);
			solution.remove(solution.size()-1);
		}

	}
	private ArrayList<ArrayList<Integer>> findSubsetsIterative(int[] num){
		ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
		ArrayList<ArrayList<Integer>> prev = new ArrayList<ArrayList<Integer>>();
		for(int i=0;i<num.length;i++){
			if(i == 0 || prev.size() == 0 || num[i] != num[i-1]){
				prev = new ArrayList<ArrayList<Integer>>();
                for(int j=0;j<res.size();j++){
                    prev.add(new ArrayList<Integer>(res.get(j)));
                }
			}
			for(ArrayList<Integer> a:prev){
				a.add(num[i]);
			}
			if(i == 0 || num[i] != num[i-1]){
                ArrayList<Integer> temp = new ArrayList<Integer>();
                temp.add(num[i]);
                prev.add(temp);
            }
            for(ArrayList<Integer> s:prev){
                res.add(new ArrayList<Integer>(s));
            }

		}
		res.add(new ArrayList<Integer>());
        return res;
	}
}